Prevent "dependency confusion" resulting in a malicious upstream gem taking precedence over a local gem

jeremy opened the pull request:
Rather than combine all gems from BOTH repositories, only include
upstream gems if a gem with the same name is NOT PRESENT LOCALLY.

This is a breaking behavior change in the interest of shipping Geminabox
in a "secure by default" configuration.

To restore the previous behavior, exposing yourself to this vuln:
* Set env var: `RUBYGEMS_PROXY_MERGE_STRATEGY=combine_local_and_remote_gem_versions`
* Add to config.ru: `Geminabox.rubygems_proxy_merge_strategy = :combine_local_and_remote_gem_versions`

One reason to keep the old behavior is if you use Geminabox to override
a public gem version with your own private gem build. In this case,
recommend using Bundler to pin to a private gem source or releasing a
forked gem with a different version instead.

https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610

simi commented on lib/geminabox.rb:
```diff
@@ -73,23 +74,24 @@ def call(env)
+    allow_replace:                  false,
+    gem_permissions:                0644,
+    rubygems_proxy:                 (ENV['RUBYGEMS_PROXY'] == 'true'),
+    rubygems_proxy_merge_strategy:  ENV['RUBYGEMS_PROXY_MERGE_STRATEGY'].fetch { :ours }.to_sym,
```
@jeremy `ENV.fetch('RUBYGEMS_PROXY_MERGE_STRATEGY') { :ours }` ?

jeremy replied:
lol yes thank you @simi

simi commented:
@sonots any chance to merge and release? Is there anything I can help with regarding that?

simi commented:
@tomlea @kerrizor @pipt please can you take a look and potentially merge and release? It is security related patch.

simi commented:
@reggieb any chance to help here?

reggieb commented on lib/geminabox/gem_list_merge.rb:
```diff
@@ -1,23 +1,38 @@
-    def self.from(*lists)
-      lists.map{|list| new(list)}.inject(:merge)
+    def self.merge(strategy, lists)
+      lists.map{|list| new(list)}.inject{|a,b| a.merge(b, strategy) }
```
This becomes a lot easier if you add `strategy` to the `#initialize` method rather than the `#merge` method.

I'd also pass the strategy in as a named parameter so that it clearly separated from the lists. It would also provide a clear place to define the expected default behaviour.
```ruby
def self.merge(*lists, strategy: :ours)
  lists.map{|list| new(strategy, list)}.inject(:merge)
end
```
Also `attr_accessor` will need to be updated to include `:strategy`

jeremy replied:
Sure; can change style to suit.

reggieb commented on lib/geminabox/gem_list_merge.rb:
```diff
@@ -1,23 +1,38 @@
 
-    def merge(other)
+    def merge(other, strategy = nil)
+      send(strategy || :ours, other)
```
Using `send` here introduces a new vulnerability as anything passed to `Geminabox.rubygems_proxy_merge_strategy` could be activated. With so few options there are better ways of doing this. For example:
```ruby
def merge(other)
  case strategy
  when :both
    both(other)
  when :theirs
    theirs(other)
  else
    ours(other)
  end
end
```

jeremy replied:
Note that `Geminabox.rubygems_proxy_merge_strategy` is not untrusted input. Would put the concern about introducing a new vuln to rest.

The motivation here is to demonstrate plugging in a named merge strategy. Can implement what matches your sense of style for this codebase.

reggieb commented on lib/geminabox/server.rb:
```diff
@@ -312,7 +312,7 @@ def remote_gem_list
 
     def combined_gem_list
-      GemListMerge.from(local_gem_list, remote_gem_list)
+      GemListMerge.new(local_gem_list).merge(remote_gem_list, Geminabox.rubygems_proxy_merge_strategy)
```
Why isn't the class method being used here? I think this should be:
```ruby
GemListMerge.merge(local_gem_list, remote_gem_list, strategy: Geminabox.rubygems_proxy_merge_strategy)
```

jeremy replied:
End result of a refactoring chain. Will match your style preference.

kerrizor commented on lib/geminabox/gem_list_merge.rb:
```diff
@@ -1,23 +1,38 @@
 
-    def self.from(*lists)
-      lists.map{|list| new(list)}.inject(:merge)
+    def self.merge(strategy, lists)
```
WDYT about allowing `strategy` to default to `nil` and use the existing merge strategy `if strategy.nil?` in order to allow for a non-breaking approach here?

jeremy replied:
Started with that. I think the vulnerability is serious enough that it merits breaking, I'm afraid. This needs to become the new default rather than requiring opt-in.

reggieb reviewed (changes requested):
The main issue I have is the use of `send` to determine which strategy method is used. I think that introduces a new vector for attack.

Also I don't like `theirs` and `ours` for the strategy names. What do those words mean in this context. Especially when there is already a concept of local and remote in the code base. I also notice `private` and `public` appear to be used as synonyms for the same concepts in the new README text. This would be a lot easier to understand if there was consistency in the terms used.

Would not strategy names like `ignore_remote_newer_versions` be more useful.

kerrizor commented on lib/geminabox/gem_list_merge.rb:
```diff
@@ -1,23 +1,38 @@
+
+    def theirs(other)
+      other.ours(self)
+    end
```
"Naming things is hard".. WDYT about `local`/`remote`?

kerrizor commented on lib/geminabox/gem_list_merge.rb:
```diff
@@ -1,23 +1,38 @@
-    def merge(other)
+    def merge(other, strategy = nil)
+      send(strategy || :ours, other)
+    end
```
We could avoid metaprogramming here and improve performance by skipping a `method_missing` lookup by using a `case/when` on `strategy` and returning a new custom exception (`UnsupportedMergeStrategy` perhaps?) We really don't have a plethora of strategies that we need to support.

jeremy replied:
Yep! Switched over to simple strategy objects responding to `#merge`.

kerrizor commented:
@reggieb heh.. great minds think alike

jeremy commented:
Thanks for the code and style feedback! I'd like feedback on the purpose and impact as well. Please do read https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610 for a sense of severity. Consider this a tracer PR: just a bit of implementation to get more traction than a bare issue.

Making a breaking behavior change is a serious decision. I encourage it, as proxying with a merged gem namespace leads directly (and trivially) to dependency confusion attack, meriting the seriousness of the change.

simi commented:
:thinking: I was actually surprised by current behaviour. I thought local gem is always preferred, and it is not related to version number.

jeremy commented:
Responding to @reggieb's review:

> The main issue I have is the use of send to determine which strategy method is used. I think that introduces a new vector for attack.

Good instincts: seeing `Object#send` in use triggers a hunt for untrusted input. There is none here, but why hunt at all?

> Also I don't like theirs and ours for the strategy names. What do those words mean in this context. Especially when there is already a concept of local and remote in the code base. I also notice private and public appear to be used as synonyms for the same concepts in the new README text. This would be a lot easier to understand if there was consistency in the terms used.

I perused the codebase and didn't pick up on that terminology. "Theirs" and "ours" are from git merge strategies. Can change to suit.

> Would not strategy names like ignore_remote_newer_versions be more useful.

I am indifferent to naming choice, so long as it's
* clear to users who don't have detailed knowledge of Geminabox internals and
* it's secure by default.

(Note that `ignore_remote_newer_versions` is not accurate, in this case.)


For this matter, having to mess with `Geminabox.rubygems_proxy_merge_strategy` at all is a red flag. What user is looking at that going "oh yeah, of course, I'll configure the strategy that I merge that proxy of my gems with." 😂

This should Just Work without needing a strategy specified. Hence the change in default and a mouthful of a config setting that's only really present for backward compatibility.

jeremy commented:
* Updated terminology to use `local` and `remote` throughout
* Embiggened `ours` and `both` strategies to `local_gems_take_precedence_over_remote_gems` and `combine_local_and_remote_gem_versions`
* Kept `local_gems_take_precedence_over_remote_gems` as default
* Refactored to a simple strategy pattern rather than pass `strategy` around as meaningful state

kerrizor commented:
@reggieb Just a heads up, I'm getting pinged about this - are you feeling ok with the PR as it is _now_? (I also don't have commit bit on this project (although I do have author rights on RubyGems..)

reggieb commented on README.md:
```diff
@@ -59,6 +59,10 @@ If you want Geminabox to carry on providing gems when rubygems.org is unavailabl
 
     Geminabox.allow_remote_failure = true
 
+By default, local gems always take precedence over upstream gems—even if the upstream gems are newer!—to prevent ["dependency confusion" supply-chain attacks](https://medium.com/@alex.birsan/dependency-confusion-4a5d60fec610) wherein an attacker pushes a remote gem with the same name as your local gem, causing you to inadvertently install a malicious gem when you `gem install` or `bundle install` ...
```
This isn't the default behaviour. This is what happens if the strategy is set to `:local_gems_take_precedence_over_remote_gems`

reggieb replied:
I'd also suggest that this section gets a heading that matches the error message. So add (before the paragraph current starting "By default, local gems always ... ):
```
### Merge Strategy
```
So that when a user gets the error "Merge strategy must be ..." they can find the relevant section of the README easily.

jeremy replied:
> This isn't the default behaviour. This is what happens if the strategy is set to :local_gems_take_precedence_over_remote_gems

This PR changes the default strategy to `:local_gems_take_precedence_over_remote_gems`.

jeremy replied:
> So add (before the paragraph currently starting "By default, local gems always ... ):

I'll leave communication and release strategy to the software maintainers. I've reverted changes to the README.

reggieb replied:
> > This isn't the default behaviour. This is what happens if the strategy is set to :local_gems_take_precedence_over_remote_gems
> 
> This PR changes the default strategy to `:local_gems_take_precedence_over_remote_gems`.

Yes you're right - I missed the setting of the default in lib/geminabox.rb.

reggieb commented on lib/geminabox/gem_list_merge.rb:
```diff
@@ -1,23 +1,38 @@
+      when :combine_local_and_remote_gem_versions
+        CombineLocalAndRemoteGemVersions
+      else
+        raise ArgumentError, "Merge strategy must be :local_gems_take_precedence_over_remote_gems (default) or :merge_local_and_remote_gem_versions"
```
If the code is going to force this setting to be set in the configuration then that should be stated in the README.

jeremy replied:
This has a default value. If you choose to set a disallowed value, it will fail.

reggieb reviewed (commented):
Looks good. My only issue is that the README needs to reflect the new behaviour in both describing the behaviour and allowing users to configure the app correctly without hitting the error first.

simi commented:
@reggieb pingie

simi commented:
@reggieb thanks for the finial review, are we waiting for @kerrizor approval now?

reggieb commented:
It's been a while since I've worked on this code base. I need to check through the release process and check I still have rights to post the updated gem. I'm busy doing my day job at the moment, so this will have to wait for a few hours.

jeremy commented:
Grateful for the encouragement, review, and release, @simi @kerrizor @reggieb! Thanks to y'all. 🙇🏼‍♂️
